crates/libm-cdylib/src/test_utils.rs

```diff
@@ -0,0 +1,130 @@
+    // This should ideally produce a link failure if libm is not dynamically
+    // linked.
+    //
+    // On MacOSX libSystem is linked (for printf, etc.) and it links libSystem_m
+    // transitively, so this doesn't help =/
+    cmd.arg("-fno-builtin")
+        .arg("-o")
+        .arg(bin_path)
+        .arg(src_path);
+    handle_err(
+        &format!("compile file: {}", src_path.display()),
+        &cmd.output().unwrap(),
+    );
+}
+
+/// Run the program and verify that it prints the expected value.
+pub(crate) fn check<T>(path: &Path, expected: T)
+where
+    T: PartialEq + std::fmt::Debug + std::str::FromStr,
+    <T as std::str::FromStr>::Err: std::fmt::Debug,
+{
+    let mut cmd = process::Command::new(path);
+
+    // Find the cdylib - we just support standard locations for now.
+    let libm_path = format!(
+        "../../target/{}/liblibm",
+        if cfg!(release_profile) {
+            "release"
+        } else {
+            "debug"
+        },
+    );
+
+    // Replace libm at runtime
+    if cfg!(target_os = "macos") {
+        // for debugging:
+        // cmd.env("DYLD_PRINT_LIBRARIES", "1");
+        // cmd.env("X", "1");
+        cmd.env("DYLD_FORCE_FLAT_NAMESPACE", "1");
+        cmd.env(
+            "DYLD_INSERT_LIBRARIES",
+            format!("{}.{}", libm_path, "dylib"),
+        );
+    } else if cfg!(target_os = "linux") {
+        cmd.env("LD_PRELOAD", format!("{}.{}", libm_path, "so"));
+    }
+    // Run the binary:
+    let output = cmd.output().unwrap();
+    handle_err(&format!("run file: {}", path.display()), &output);
+    // Parse the result:
+    let result = String::from_utf8(output.stdout.clone())
+        .unwrap()
+        .parse::<T>();
+
+    if result.is_err() {
+        panic!(format_output("check (parse failure)", &output));
+    }
+    // Check the result:
+    let result = result.unwrap();
+    assert_eq!(result, expected, "{}", format_output("check", &output));
+}
+
+pub(crate) fn handle_err(step: &str, output: &process::Output) {
+    if !output.status.success() {
+        eprintln!("{}", format_output(step, output));
+        panic!();
+    }
+}
+
+pub(crate) fn format_output(
+    step: &str,
+    process::Output {
+        status,
+        stdout,
+        stderr,
+    }: &process::Output,
+) -> String {
+    let mut s = format!("\nFAILED[{}]: exit code {:?}\n", step, status.code());
+    s += &format!(
+        "FAILED[{}]: stdout:\n\n{}\n\n",
+        step,
+        String::from_utf8(stdout.to_vec()).unwrap()
+    );
+    s += &format!(
+        "FAILED[{}]: stderr:\n\n{}\n\n",
+        step,
+        String::from_utf8(stderr.to_vec()).unwrap()
+    );
+    s
+}
+
+/// For a given Rust type `x`, this prints the name of the type in C,
+/// as well as the printf format specifier used to print values of that type.
+pub(crate) fn ctype_and_printf_format_specifier(x: &str) -> (&str, &str) {
+    match x {
+        "f32" => ("float", "%f"),
+        "f64" => ("double", "%f"),
```

Comment:
should this be `%lf`?

Reply:
IIRC, %f prints doubles. There is no format string for printing floats AFAIK - %f converts floats into a double, and prints that.

Reply:
Would it make sense when `%f` is a double anyway? maybe `%.24f` for floats `%.54f` for doubles makes more sense?

Reply:
Or even better maybe just `e`

Reply:
For the linking tests it doesn't really matter. I'll add a comment about this.

Reply:
So I've added a comment explaining what this does right now. 

If we wanted to verify the results of this lib, I think I would prefer doing so as part of the libm-test system. We could dynamically link libm as the system library, and just verify that the results of the C ABI match the results that are produced by the Rust library, or something like that.

The problem which checking the ABI here is that, if for some reason linking failed, or the symbol has a slightly different name, etc. the functions from the system library are invoked. So if we were to just simply verify the results, we can't tell whether its our libm or the system libm that's being invoked.

Reply:
Maybe there is a way to reliably prevent libm from being linked on Linux, while still being able to link, e.g., `fprintf` and other parts of the standard library. On MacOSX, this is IMO not worth it - everything links `libSystem.dylib`, which always dynamically links `libm`.